rfd/0185-k8s-access-non-cert-routing.md

```diff
@@ -0,0 +1,247 @@
+`KubernetesCluster` to produce a certificate explicitly for access to
+Kubernetes. When producing a kubeconfig, the client should use the user's 
+standard identity.
+
+To configure Kubernetes clients to work with this, the Kubeconfig generated by
+Teleport would be updated to include this base path within the
+`clusters[name].cluster.server` field:
+
+```yaml
+apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority-data: ...omitted... # No-change
+    # Before:
+    # server: https://example.teleport.sh/
+    # After:
+    server: https://example.teleport.sh/v1/teleport/base64-teleport-cluster-name/base64-k8s-cluster-name
+    tls-server-name: kube-teleport-proxy-alpn.example.teleport.sh # No-change
+  name: example-example
+```
+
+#### Backwards/Forwards Compatibility
+
+By retaining the existing behaviour and endpoints within the Kube Forwarder,
+older clients using a kubeconfig and certificate generated with the
+`KubernetesCluster` attribute will continue to function.
+
+- VX.Y.0:
+  - Introduce new Kube Forwarder behaviour.
+  - Introduce opt-in `kubernetes/v2` `tbot` output. This will allow
+    users to opt-in to the new behaviour once they are satisfied that their
+    Auth Server, Proxy and Kubernetes Agents have been upgraded to the required
+    version.
+- V(X+1).0.0:
+  - Switch `tsh` to new behaviour. No longer requests the signing of a
+    certificate with the `KubernetesCluster` attribute and uses standard user
+    identity to authenticate.
+  - Switch `tbot` `kubernetes/v1` output to use `kubernetes/v2` implementation
+    under the hood.
+- V(X+2).0.0:
+  - It is now "safe" to remove the old attribute based routing from the Kube
+    Forwarder. We may wish to delay this further if we have concerns about
+    older client compatability.
+
+#### Client Compatability
+
+Kubernetes API clients have been known to vary greatly in their support for 
+more unusual configuration. We have witnessed this in the past when it came to
+supporting the ability to explicitly provide the SNI to use when making
+requests.
+
+Therefore, we should ensure that Kubernetes clients and SDKs do support
+providing a base path as part of the server address (e.g they do not strip off
+the path and only use the hostname/port).
+
+The key customer we are working with has specified that they use the following
+Kubernetes clients:
+
+- `kubectl`
+- Python Kubernetes SDK
+
+Both of these clients correctly handle a base-path as configured in Kubeconfig:
+
+```yaml
+clusters:
+  - cluster:
+      server: http://127.0.0.1:8888/example/base/path
+    name: test-base-path
+```
+
+#### Analytics
+
+It would no longer be possible to track usage of Kubernetes Access for analytics
+purposes via the certificate generation event. However, we have other events
+that indicate Kubernetes Access usage (e.g an event per request), that are
+better suited for this purpose - so this seems acceptable.
```

Comment:
Note that when `kubectl` performs discovery, it requests a single certificate but retrieves all API resource definitions. In theory, this could generate duplicate events, but Teleport disregards discovery requests for usage purposes, so nothing will change neither the metrics will be duplicated.

As for billing, since we count Kubernetes requests, this behavior won’t affect anything.